core/src/compile/mod.rs

```diff
@@ -507,460 +48,210 @@ impl<'a> NamespaceCompiler<'a> {
+                    .into_iter()
+                    .chain(pushbacks.into_iter())
+                    .chain(descents.into_iter())
+                    .collect::<Vec<_>>()
+            };
 
-            let own_dependencies: Vec<Scope> = scope
-                .ordering
-                .iter()
-                .filter(|attr| !scope.children.get(*attr).unwrap().compiled.is_built())
-                .map(|attr| next_scope.clone().into_at(attr))
-                .collect();
-            stage_2!(
-                "{}: {} dependencies left",
-                next_scope,
-                own_dependencies.len()
-            );
-
-            let mut foreign_dependencies = Vec::new();
-            for reference in refs.iter() {
-                if !self.vtable.is_built(reference, &next_scope)? {
-                    foreign_dependencies.push(reference.clone());
+            if !next.is_empty() {
+                eprintln!("dependencies not satisfied: {:?}", next);
+                if matches!(self.state.project_mut(address.clone())?.output_mut().waiting(), OutputState::Waiting) {
+                    return Err(anyhow!("cycle detected"));
                 }
-            }
-            stage_2!(
-                "{}: {} references left",
-                next_scope,
-                foreign_dependencies.len()
-            );
-
-            if !own_dependencies.is_empty() || !foreign_dependencies.is_empty() {
-                visits.push_back(next_scope);
-                visits.extend(own_dependencies.into_iter().chain(foreign_dependencies));
+                visits.push(address);
+                visits.extend(next);
                 continue;
             }
 
-            let mut drivers = BTreeMap::new();
-
-            let ctx = self
-                .vtable
-                .get(&next_scope)
-                .map(|locals| {
-                    let keys: Vec<_> = locals.locals.keys().map(|key| key.to_string()).collect();
-                    keys.as_slice().join(", ")
-                })
-                .unwrap_or_else(|_| "{empty}".to_string());
-            stage_2!("{}: context extends to {}", next_scope, ctx);
+            if !address.is_root() {
+                let mut parent = address.clone();
+                let this = parent.shallower().unwrap();
+                // It is not necessary to order nodes that are not bound to be wrapped in `Ordered`
+                if self.vtable.contains(&parent) {
+                    self.state.project_mut(parent)?.scope_mut().push(this).unwrap();
+                }
+            }
 
+            let state = self.state.project_mut(address.clone())?;
+            let vtable = &mut self.vtable;
+            let mut children = BTreeMap::new();
             let content_compiler = ContentCompiler {
-                scope: next_scope.clone(),
-                cursor: next,
-                drivers: &mut drivers,
-                vtable: &mut self.vtable,
+                scope: address.clone(),
```

Comment:
Is `scope` still the right field name here? (Since I think you changed the `Scope` struct to be `Address`.